src/core/cri/cri.ts

```diff
@@ -0,0 +1,163 @@
+/**
+ * cri.ts
+ *
+ * Chrome Remote Interface
+ *
+ * Copyright (c) 2022 by RStudio, PBC.
+ */
+
+import cdp from "./deno-cri/index.js";
+import { decode } from "encoding/base64.ts";
+import { getBrowserExecutablePath } from "../puppeteer.ts";
+import { Semaphore } from "../lib/semaphore.ts";
+
+async function waitForServer(port: number, timeout = 3000) {
+  const interval = 500;
+  let soFar = 0;
+
+  do {
+    try {
+      const response = await fetch(`http://localhost:${port}/json/list`);
+      if (response.status !== 200) {
+        soFar += interval;
+        await new Promise((resolve) => setTimeout(resolve, interval));
+        continue;
+      } else {
+        return true;
+      }
+    } catch (_e) {
+      soFar += interval;
+      await new Promise((resolve) => setTimeout(resolve, interval));
+    }
+  } while (soFar < timeout);
+  return false;
+}
+
+const criSemaphore = new Semaphore(1);
+
+export function withCriClient<T>(
+  fn: (client: Awaited<ReturnType<typeof criClient>>) => Promise<T>,
+  appPath?: string,
+  port = 9222,
+): Promise<T> {
+  return criSemaphore.runExclusive(async () => {
+    const client = await criClient(appPath, port);
+    try {
+      const result = await fn(client);
+      await client.close();
+      return result;
+    } catch (e) {
+      await client.close();
+      throw e;
+    }
+  });
+}
+
+export async function criClient(appPath?: string, port = 9222) {
```

Comment:
Do we need to find a new open port here? (or gracefully handle finding a different port if this one is taken?). We have some tools for this in core

Reply:
Done.